server/src/main/java/org/apache/druid/segment/realtime/appenderator/SegmentsAndCommitMetadata.java

```diff
@@ -20,28 +20,50 @@
 package org.apache.druid.segment.realtime.appenderator;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
 import org.apache.druid.segment.SegmentUtils;
 import org.apache.druid.timeline.DataSegment;
 
 import javax.annotation.Nullable;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
+import java.util.Set;
 
 public class SegmentsAndCommitMetadata
 {
   private static final SegmentsAndCommitMetadata NIL = new SegmentsAndCommitMetadata(Collections.emptyList(), null);
 
   private final Object commitMetadata;
   private final ImmutableList<DataSegment> segments;
+  private final ImmutableSet<DataSegment> upgradedSegments;
```

Comment:
Why does this class need to distinguish between upgraded segments and root ones? From the POV of this class, all of these segments were committed and thus should all be handed off.

Reply:
From the POV of segment handoff, the original and upgraded segments need not be distinguished from each other.

However, there are sanity checks where we must ensure that the committed segments are the same as the ones that were requested. Maintaining the upgraded segments separately helps with this check.
When the check fails and segments need to be killed from deep storage, we would also have to ensure that we do not run into errors trying to kill the same deep storage location multiple times. While this could be done by creating a set of load specs before every kill, I think it may be neater to maintain the orignal and upgraded sets separately for such purposes.

Reply:
Could you please link the places where these sanity checks are being performed?

Reply:
Looking at the code, the cleanup happens in `BaseAppenderatorDriver.publishInBackground` itself. For cleanup, we would use the original requested `segmentsAndCommitMetadata` and not the one we freshly created and returned.
That's what we seem to be doing in the current impl in this PR as well.